services/filter.service.js

```diff
@@ -0,0 +1,192 @@
+    }
+    return false;
+  }
+
+  function getValueByFieldType(val, keyname, listfields) {
+    var type = getTypeOf(keyname, listfields);
+    if (type == 'NUMERIC' || type == 'PERCENT' || type == 'AMOUNT')
+      return Number(val);
+    else {
+      return val;
+    }
+  }
+
+  function getConditionAfter(val, obj, listfields) {
+
+    var jsonRes = {};
+    Object.keys(obj).forEach(function (key) {
+      if (key[0] === '$') {
+        if (CONDITIONAL_OPERATORS.indexOf(key) != -1) {
+          jsonRes = getConditionBefore(val, obj, listfields);
+        } else if (tryKeyByType(key, val, listfields)) {
+          jsonRes[key] = getValueByFieldType(obj[key], val, listfields);
+        }
+      }
+      else {
+        return undefined;
+      }
+    });
+    return (jsonRes);
+  }
+
+  function getConditionBefore(currentKey, array, listfields) {
+    var jsontab = [];
+    if (Array.isArray(array) === true) {
+      for(var i = 0; i < array.length; i++) {
+        var value = array[i];
+        if (typeof value !== 'object') {
+          value = {'$eq': value};
+        }
+        jsontab.push(getConditionAfter(currentKey, value, listfields));
+      }
+    } else {
+      Object.keys(array).forEach(function (key) {
+        var jsonRes = {};
+        if (CONDITIONAL_OPERATORS.indexOf(key) != -1) {
+          jsonRes[key] = getConditionBefore(currentKey, array[key], listfields);
+          if (jsonRes[key] === undefined || !tryKeyByType(key, jsonRes[key], listfields)) {
+            return undefined;
+          }
+          if (currentKey !== null) {
+            jsontab = jsonRes;
+          }
+          else
+            jsontab.push(jsonRes);
+        }
+        else if (key[0] !== '$') {
+          var kval = getValIfValExist(key, listfields);
+          if (kval === undefined)
+            return (undefined);
+          if (typeof array[kval] !== 'object') {
+            array[kval] = {'$eq': array[kval]};
+          }
+          if (Array.isArray(array[kval]) === true) {
+            for (var i = 0; i < array[kval].length; i++) {
+              var value = array[kval][i];
+              if (typeof value !== 'object') {
+                value = {'$eq': value};
+              }
+              var jsonRes = {};
+              jsonRes[kval] = getConditionAfter(kval, value, listfields);
+              jsontab.push(jsonRes);
+            }
+          } else {
+            jsonRes[kval] = getConditionAfter(kval, array[kval], listfields);
+            jsontab.push(jsonRes);
+          }
+        }
+        else {
+          var jsonRes2 = {};
+          jsonRes2[key] = array[key];
+          jsontab.push(getConditionAfter(currentKey, jsonRes2, listfields));
+        }
+      });
+    }
+    return jsontab;
+  }
+
+  this.buildMongoQuery = function (jsonRes, filters, listfields) {
+    if (typeof filters !== 'undefined') {
+      //var jsonVal = {};
+      if ((jsonRes["$and"] = getConditionBefore(null, filters, listfields)) === undefined)
```

Comment:
Après réflexion, il faut aussi que le `jsonRes` par défaut soit inclus dans ce `$and`.